Apply bigdecimal feature (solved issue)

devgony opened the pull request:
# Apply bigdecimal feature (#337) 

1. Add bigdecimal
```rs
// Cargo.toml
sqlparser = { version = "0.11", features = ["serde", "bigdecimal"] }
bigdecimal = { version = "0.3", features = ["serde"]}
```
2. Add serde_str to solve `DeserializeAnyNotSupported` error.
```rs
// Cargo.toml
serde_str="0.1.0"
```
```rs
// src/ast/ast_literal.rs
pub enum AstLiteral {
    Boolean(bool),
    #[serde(with = "serde_str")]
    Number(BigDecimal),
...
```

3. Bigdecimal has larger floating point than f64 
    => fixed with rectified float.
```rs
// sqrt_power.rs
            "SELECT POWER(32,0.3) as power_with_float FROM SingleItem",
            Ok(select!(
                power_with_float
                F64;
                f64::from(32).powf(0.3) // expected: [F64(2.82842712474619)], found: [F64(2.8284271247461907)
            )),
```

devgony committed:
enable feature bigdecimal

devgony committed:
try bigdecimal add trait

devgony committed:
should fix floating point

devgony committed:
edit bigdecimal test

devgony committed:
rebase on latest main, handle dividedByZero

devgony committed:
add serde_str

devgony committed:
1. fix concat test with rectified float
2. fix sqrt_power test with rectified float
3. handle unwrap with UnreachableNumberParsing

devgony committed:
add cfg(feature=bigdecimal) for clippy

devgony committed:
enble bigdecimal always

devgony committed:
use Cow with bigdecimal

panarch reviewed (changes requested):
Sorry for the late review.
Looks nice :)
Let's just try to remove unnecessary `unwrap()`s.

devgony committed:
remove unwrap

devgony committed:
enable string-only

devgony committed:
remove redundant Neg, println

panarch commented:
We can now replace `bigdecimal::ToPrimitive` to `data::bigdecimal_ext`.
Then, we can remove `BigDecimal::is_integer` checks.

```rust
use bigdecimal::BigDecimal;                               
                                                          
pub trait BigDecimalExt {                                 
    fn to_i64(&self) -> Option<i64>;                      
                                                          
    fn to_f64(&self) -> Option<f64>;                      
}                                                         
                                                          
impl BigDecimalExt for BigDecimal {                       
    fn to_i64(&self) -> Option<i64> {                     
        match self.is_integer() {                         
            true => bigdecimal::ToPrimitive::to_i64(self),
            false => None,                                
        }                                                 
    }                                                     
                                                          
    fn to_f64(&self) -> Option<f64> {                     
        bigdecimal::ToPrimitive::to_f64(self)             
    }                                                     
}                                                         
```

panarch commented on src/tests/function/cast.rs:
```diff
@@ -23,7 +23,7 @@ test_case!(cast_literal, async move {
         ),
         (
             r#"SELECT CAST(3 AS BOOLEAN) AS cast FROM Item"#,
-            Err(ValueError::LiteralCastToBooleanFailed("3".to_owned()).into()),
+            Ok(select!(cast Bool; true)),
```
This issue is not supposed to change the `CAST` behavior.
`CAST` from `Value::Str` to `Value::Bool` only accept `1` as `TRUE`.
Let's keep the original behavior.

devgony committed:
implement BigDecimalExt

devgony committed:
remove is_integer

devgony committed:
1 is the only true value

devgony committed:
recover correct test case

devgony committed:
prettify test case

devgony committed:
fix primative::to_i64 to BigDeciamlExt::to_i64

devgony committed:
prettify ValueErrors
remove LiteralCastFromNumberToIntegerFailed
change LiteralCastFromNumberToFloatFailed => UnreachableLiteralCastFromNumberToFloat

panarch reviewed (approved):
Beautiful, everything looks good!
Thanks a lot 👍 👍 👍
